### PR TITLE
CLN: fix build warning in c_timestamp.pyx

### PR DESCRIPTION
**Pull request by ghost**
```
pandas/_libs/tslibs/c_timestamp.c: In function ‘__pyx_f_6pandas_5_libs_6tslibs_11c_timestamp_10_Timestamp__get_start_end_field’:
pandas/_libs/tslibs/c_timestamp.c:7257:25: warning: ‘__pyx_t_10’ may be used uninitialized in this function [-Wmaybe-uninitialized]
 7257 |   __pyx_t_5numpy_int8_t __pyx_t_10[1];
      |  
```

Retrying fix dropped from https://github.com/pandas-dev/pandas/pull/27371

a review [comment](https://github.com/pandas-dev/pandas/pull/27371#discussion_r303179721) there disliked reintroducing `ndarray` here because the file had been converted to using memoryviews. `get_start_end_field` returns a view of type `npy_bool` so`out` should therefore be either a memoryview or an ndarray, and bool memoryviews will only be supported in cython 0.30
(see cython/cython/pull/2676).

### CODE REVIEW

**Comment by WillAyd** [comment]
Haven't closely followed that conversation or checked links so apologies if stupid question but does a `unit8` memory view not work in the meantime? IIRC we've used that in the past instead

**Comment by pilkibun** [comment]
Not stupid at all. But no, that doesn't work either because on <-0.30  bool arrays require an explicit cast even for assigning to a npy_bool array. What you get otherwise is

```
ValueError: Does not understand character buffer dtype format string ('?')
```

The linked cython PR fixes this in the next release.

**Comment by jbrockmendel** [comment]
It looks like the linked cython issue makes it unnecessary to specify `cast=True` in some ndarray declarations.  We use that in a few places, so once we update to require cython>=0.30.0, feel free to take point on removing that no-longer-necessary usage

**Comment by pilkibun** [comment]
ok. There's also the conditional gil support in 0.30 which is needed. Should probably update soon after release.

File: pandas/_libs/tslibs/c_timestamp.pyx [hunk 1]
```diff
@@ -320,7 +320,7 @@ cdef class _Timestamp(datetime):
         cdef:
             int64_t val
             dict kwds
```
Comment (jreback) [inline pandas/_libs/tslibs/c_timestamp.pyx]: we type these as uint_8

i don’t believe we use npy_bool anywhere and would be a consistently change

**Comment by pilkibun** [comment]
ping

File: pandas/_libs/tslibs/c_timestamp.pyx [hunk 2]
```diff
@@ -320,7 +320,7 @@ cdef class _Timestamp(datetime):
...
-            int8_t out[1]
+            ndarray[uint8_t, cast=True] out
```
Comment (jreback) [inline pandas/_libs/tslibs/c_timestamp.pyx]: does this actually need the cast arg? we don't use this anywhere else IIRC

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/tslibs/c_timestamp.pyx (hunk 2 above)
Comment (pilkibun) [reply pandas/_libs/tslibs/c_timestamp.pyx]: please read the issue.

Comment (jreback) [reply pandas/_libs/tslibs/c_timestamp.pyx]: we are not requiring cython >= 0.30 anytime soon, but I see for bools this is needed, ok

**Review by jreback — APPROVED** [review]

File: pandas/_libs/tslibs/c_timestamp.pyx (hunk 2 above)
Comment (pilkibun) [reply pandas/_libs/tslibs/c_timestamp.pyx]: thanks.

**Pull request merged**